File: src/lexpy.py
# Libaries
import re;



# Generator function
def gen(lex):
	# Language descriptor
	descriptor = lex.Descriptor([
		u"IGNORE",
		u"MEMBER",
		u"NEXT_IS_MEMBER",
		u"NEXT_NO_OP_PREFIX",
		u"START_STRING",
		u"STRING_TRIPLE",
		u"START_COMMENT",
		u"BRACKET",
		u"BRACKET_CLOSE",
	], u"IGNORE");
	flags = descriptor.flags;
	descriptor.define_types({
		u"INVALID": 0,
		u"KEYWORD": 0,
		u"IDENTIFIER": flags.NEXT_NO_OP_PREFIX,
		u"NUMBER": flags.NEXT_NO_OP_PREFIX,
		u"STRING": flags.NEXT_NO_OP_PREFIX,
		u"OPERATOR": 0,
		u"DECORATOR": 0,
		u"WHITESPACE": flags.IGNORE,
		u"COMMENT": flags.IGNORE,
	});
	keywords = {
		u"and": 0,
		u"as": 0,
		u"assert": 0,
		u"break": 0,
		u"class": 0,
		u"continue": 0,
		u"def": 0,
		u"del": 0,
		u"elif": 0,
		u"else": 0,
		u"except": 0,
		u"finally": 0,
		u"for": 0,
		u"from": 0,
		u"global": 0,
		u"if": 0,
		u"import": 0,
		u"in": 0,
		u"is": 0,
		u"lambda": 0,
		u"nonlocal": 0,
		u"not": 0,
		u"or": 0,
		u"pass": 0,
		u"raise": 0,
		u"return": 0,
		u"try": 0,
		u"while": 0,
		u"with": 0,
		u"yield": 0,
	};
	operators = lex.tree({
		u">>=": 0,
		u">>": 0,
		u">=": 0,
		u">": 0,

		u"<<=": 0,
		u"<<": 0,
		u"<=": 0,
		u"<": 0,
		u"<>": 0,

		u"==": 0,
		u"=": 0,

		u"!=": 0,

		u"&=": 0,
		u"&": 0,

		u"|=": 0,
		u"|": 0,

		u"+=": 0,
		u"+": 0,

		u"-=": 0,
		u"-": 0,

		u"**=": 0,
		u"**": 0,
		u"*=": 0,
		u"*": 0,

		u"//=": 0,
		u"//": 0,
		u"/=": 0,
		u"/": 0,

		u"%=": 0,
		u"%": 0,

		u"^=": 0,
		u"^": 0,

		u"~": 0,
		u":": 0,
		u";": 0,
		u",": 0,
		u".": flags.NEXT_IS_MEMBER | flags.NEXT_NO_OP_PREFIX,
		u"...": 0,

		u"\\": 0,

		u"(": flags.BRACKET,
		u"[": flags.BRACKET,
		u"{": flags.BRACKET,
		u")": flags.BRACKET | flags.BRACKET_CLOSE | flags.NEXT_NO_OP_PREFIX,
		u"]": flags.BRACKET | flags.BRACKET_CLOSE | flags.NEXT_NO_OP_PREFIX,
		u"}": flags.BRACKET | flags.BRACKET_CLOSE,

		u"#": flags.START_COMMENT,
		u"\"": flags.START_STRING,
		u"\"\"\"": flags.START_STRING | flags.STRING_TRIPLE,
		u"\'": flags.START_STRING,
		u"\'\'\'": flags.START_STRING | flags.STRING_TRIPLE,
	});

	# Matching logic
	re_comment = re.compile(u"[^\\r\\n\u2028\u2029]*");
	re_newlines_search = re.compile(u"[\\r\\n\u2028\u2029]");
	re_newlines_split = re.compile(u"[\\n\u2028\u2029]|\\r\\n?");

	def match_string(self, t_info):
		# Match to end of string
		escaped = False;
		p = t_info[2];
		p_max = len(self.text);
		quote = self.text[self.pos]
		quote_count = 0;
		quote_length = 1;
		if ((t_info[1] & self.descriptor.flags.STRING_TRIPLE) != 0): quote_length = 3;

		while (p < p_max):
			c = self.text[p];
			if (escaped):
				escaped = False;
				if (c == u"\r" and p + 1 < p_max and self.text[p + 1] == u"\n"):
					p += 1;
			else:
				if (c == quote):
					quote_count += 1;
					if (quote_count >= quote_length):
						p += 1;
						break;
				else:
					quote_count = 0;
					if (c == u"\\"):
						escaped = True;
					elif (string_contains_newline(c) and quote_length == 1):
						break;

			p += 1;

		t_info[0] = self.descriptor.STRING;
		t_info[2] = p;

	def match_comment(self, t_info):
		# Match the comment
		p = t_info[2];
		m = re_comment.match(self.text, p);

		# Create the token
		t_info[0] = self.descriptor.COMMENT;
		t_info[2] = m.end();

	# Token creation functions
	def create_token_number(self, flags, p):
		if ((self.previous.flags & self.descriptor.flags.NEXT_NO_OP_PREFIX) == 0 or self.match_tree(operators, self.pos, self.pos + 1) is None):
			return self.create_token(descriptor.NUMBER, flags, p);

		return None;

	def create_token_word(self, flags, p):
		token_type = self.descriptor.IDENTIFIER;

		if ((self.previous.flags & self.descriptor.flags.NEXT_IS_MEMBER) != 0):
			# Member
			flags |= self.descriptor.flags.MEMBER;

		# Check if keyword
		word = self.text[self.pos : p];
		if (word in keywords):
			token_type = self.descriptor.KEYWORD;
			flags |= keywords[word];

		return self.create_token(token_type, flags, p);

	def create_token_operator(self, flags, p):
		t_info = [ self.descriptor.OPERATOR , flags , p ];

		if ((flags & self.descriptor.flags.START_STRING) != 0):
			# String
			match_string(self, t_info);
		elif ((flags & self.descriptor.flags.START_COMMENT) != 0):
			# Comment
			match_comment(self, t_info);

		return self.create_token(t_info[0], t_info[1], t_info[2]);

	# Create descriptor
	descriptor.define_state_names([ u"DEFAULT" ]);
	descriptor.define_state([ # state 0
		[
			lex.check_regex(u"\\s+"), # whitespace
			lex.create_token(descriptor.WHITESPACE),
		],
		[
			lex.check_regex(u"[+-]?(?:0[xX](?:[0-9a-fA-F]+)|[0-9]+(?:\\.[0-9]*)?(?:[eE][+-]?[0-9]+)?|\\.[0-9]+(?:[eE][+-]?[0-9]+)?)"), # number
			create_token_number,
		],
		[
			lex.check_regex(u"[\\w]+"), # word
			create_token_word,
		],
		[
			lex.check_regex(u"@[\\w]+"), # decorator
			lex.create_token(descriptor.DECORATOR),
		],
		[
			lex.check_tree(operators), # operator
			create_token_operator,
		],
		[
			lex.check_regex(u"[^\\s\\w@" + re.escape(lex.to_regex_class(operators)) + u"]+"), # invalid
			lex.create_token(descriptor.INVALID),
		],
	], 0);

	# Additional functions
	def string_contains_newline(text):
		return re_newlines_search.search(text) is not None;

	def string_splitlines(text):
		return re_newlines_split.split(text);

	setattr(descriptor, u"string_contains_newline", string_contains_newline);
	setattr(descriptor, u"string_splitlines", string_splitlines);

	# Complete
	return descriptor;

File: src/lexpy.js
var lexpy = (function () {
	"use strict";

	// Basic environment derection; can be hardcoded if the environment is known
	var module = (function (module) {
		try {
			if (window) return {};
		}
		catch (e) {}

		return module;
	})(this);



	// Generator function
	module.gen = function (lex) {

		// Language descriptor
		var descriptor = new lex.Descriptor([ //{
			"IGNORE",
			"MEMBER",
			"NEXT_IS_MEMBER",
			"NEXT_NO_OP_PREFIX",
			"START_STRING",
			"STRING_TRIPLE",
			"START_COMMENT",
			"BRACKET",
			"BRACKET_CLOSE",
		], "IGNORE"); //}
		var flags = descriptor.flags;
		descriptor.define_types({
			"INVALID": 0,
			"KEYWORD": 0,
			"IDENTIFIER": flags.NEXT_NO_OP_PREFIX,
			"NUMBER": flags.NEXT_NO_OP_PREFIX,
			"STRING": flags.NEXT_NO_OP_PREFIX,
			"OPERATOR": 0,
			"DECORATOR": 0,
			"WHITESPACE": flags.IGNORE,
			"COMMENT": flags.IGNORE,
		});
		var keywords = {
			"and": 0,
			"as": 0,
			"assert": 0,
			"break": 0,
			"class": 0,
			"continue": 0,
			"def": 0,
			"del": 0,
			"elif": 0,
			"else": 0,
			"except": 0,
			"finally": 0,
			"for": 0,
			"from": 0,
			"global": 0,
			"if": 0,
			"import": 0,
			"in": 0,
			"is": 0,
			"lambda": 0,
			"nonlocal": 0,
			"not": 0,
			"or": 0,
			"pass": 0,
			"raise": 0,
			"return": 0,
			"try": 0,
			"while": 0,
			"with": 0,
			"yield": 0,
		};
		var operators = lex.tree({
			">>=": 0,
			">>": 0,
			">=": 0,
			">": 0,

			"<<=": 0,
			"<<": 0,
			"<=": 0,
			"<": 0,
			"<>": 0,

			"==": 0,
			"=": 0,

			"!=": 0,

			"&=": 0,
			"&": 0,

			"|=": 0,
			"|": 0,

			"+=": 0,
			"+": 0,

			"-=": 0,
			"-": 0,

			"**=": 0,
			"**": 0,
			"*=": 0,
			"*": 0,

			"//=": 0,
			"//": 0,
			"/=": 0,
			"/": 0,

			"%=": 0,
			"%": 0,

			"^=": 0,
			"^": 0,

			"~": 0,
			":": 0,
			";": 0,
			",": 0,
			".": flags.NEXT_IS_MEMBER | flags.NEXT_NO_OP_PREFIX,
			"...": 0,

			"\\": 0,

			"(": flags.BRACKET,
			"[": flags.BRACKET,
			"{": flags.BRACKET,
			")": flags.BRACKET | flags.BRACKET_CLOSE | flags.NEXT_NO_OP_PREFIX,
			"]": flags.BRACKET | flags.BRACKET_CLOSE | flags.NEXT_NO_OP_PREFIX,
			"}": flags.BRACKET | flags.BRACKET_CLOSE,

			"#": flags.START_COMMENT,
			"\"": flags.START_STRING,
			"\"\"\"": flags.START_STRING | flags.STRING_TRIPLE,
			"\'": flags.START_STRING,
			"\'\'\'": flags.START_STRING | flags.STRING_TRIPLE,
		});

		// Matching logic
		var re_comment = /[^\r\n\u2028\u2029]*/g,
			re_newlines_search = /[\r\n\u2028\u2029]/,
			re_newlines_split = /[\n\u2028\u2029]|\r\n?/g;

		var match_string = function (t_info) {
			// Match to end of string
			var escaped = false,
				p = t_info[2],
				p_max = this.text.length,
				quote = this.text[this.pos],
				quote_count = 0,
				quote_length = ((t_info[1] & this.descriptor.flags.STRING_TRIPLE) !== 0) ? 3 : 1,
				c;

			for (; p < p_max; ++p) {
				c = this.text[p];
				if (escaped) {
					escaped = false;
					if (c === "\r" && p + 1 < p_max && this.text[p + 1] === "\n") {
						++p;
					}
				}
				else {
					if (c === quote) {
						if (++quote_count >= quote_length) {
							++p;
							break;
						}
					}
					else {
						quote_count = 0;
						if (c === "\\") {
							escaped = true;
						}
						else if (this.descriptor.string_contains_newline(c) && quote_length === 1) {
							break;
						}
					}
				}
			}

			t_info[0] = this.descriptor.STRING;
			t_info[2] = p;
		};
		var match_comment = function (t_info) {
			// Check which type
			var p = t_info[2],
				re = re_comment,
				m;

			// Match the comment
			re.lastIndex = p;
			m = re.exec(this.text);

			// Create the token
			t_info[0] = this.descriptor.COMMENT;
			t_info[2] = m.index + m[0].length;
		};

		// Checks/states
		descriptor.define_state_names([ "DEFAULT" ]);
		descriptor.define_state([ //{ state 0
			[
				lex.check_regex("\\s+"), // whitespace
				lex.create_token(descriptor.WHITESPACE),
			],
			[
				lex.check_regex("[+-]?(?:0[xX](?:[0-9a-fA-F]+)|[0-9]+(?:\\.[0-9]*)?(?:[eE][+-]?[0-9]+)?|\\.[0-9]+(?:[eE][+-]?[0-9]+)?)"), // number
				function (flags, p) {
					if ((this.previous.flags & this.descriptor.flags.NEXT_NO_OP_PREFIX) === 0 || this.match_tree(operators, this.pos, this.pos + 1) === null) {
						return this.create_token(descriptor.NUMBER, flags, p);
					}
					return null;
				},
			],
			[
				lex.check_regex("[\\w]+"), // word
				function (flags, p) {
					var token_type = this.descriptor.IDENTIFIER,
						word = this.text.substr(this.pos, p - this.pos);

					if ((this.previous.flags & this.descriptor.flags.NEXT_IS_MEMBER) !== 0) {
						// Member
						flags |= this.descriptor.flags.MEMBER;
					}

					// Check if keyword
					if (Object.prototype.hasOwnProperty.call(keywords, word)) {
						token_type = this.descriptor.KEYWORD;
						flags |= keywords[word];
					}

					return this.create_token(token_type, flags, p);
				},
			],
			[
				lex.check_regex("@[\\w]+"), // decorator
				lex.create_token(descriptor.DECORATOR),
			],
			[
				lex.check_tree(operators), // operator
				function (flags, p) {
					var t_info = [ this.descriptor.OPERATOR , flags , p ];

					if ((flags & this.descriptor.flags.START_STRING) !== 0) {
						// String
						match_string.call(this, t_info);
					}
					else if ((flags & this.descriptor.flags.START_COMMENT) !== 0) {
						// Comment
						match_comment.call(this, t_info);
					}
					else if ((flags & this.descriptor.flags.START_REGEX) !== 0) {
						// Regex
						match_regex.call(this, t_info);
					}

					return this.create_token(t_info[0], t_info[1], t_info[2]);
				},
			],
			[
				lex.check_regex("[^\\s\\w@" + lex.regex_escape(lex.to_regex_class(operators)) + "]+"), // invalid
				lex.create_token(descriptor.INVALID),
			],
		], 0); //}

		// Additional functions
		descriptor.string_contains_newline = function (text) {
			return re_newlines_search.test(text);
		};
		descriptor.string_splitlines = function (text) {
			var parts = [],
				start = 0,
				m;

			re_newlines_split.lastIndex = 0;
			while ((m = re_newlines_split.exec(text)) !== null) {
				parts.push(text.substr(start, m.index - start));
				start = m.index + m[0].length;
			}

			parts.push(text.substr(start));

			return parts;
		};

		// Complete
		return descriptor;

	};



	// Complete
	return module;

}).call(this);


